sdk/core/azure-core/azure/core/credentials.py

```diff
@@ -28,3 +29,45 @@ def get_token(self, *scopes, **kwargs):
     from collections import namedtuple
 
     AccessToken = namedtuple("AccessToken", ["token", "expires_on"])
+
+__all__ = ["AzureKeyCredential"]
+
+
+class AzureKeyCredential(object):
+    """Credential type used for authenticating to an Azure service.
+    It provides the ability to update the key without creating a new client.
+
+    :param str key: The key used to authenticate to an Azure service
+    :raises: TypeError
+    """
+
+    def __init__(self, key):
+        # type: (str) -> None
+        if not isinstance(key, six.string_types):
+            raise TypeError("key must be a string.")
+        self._key = key  # type: str
+
+    @property
+    def key(self):
+        # type () -> str
+        """The value of the configured key.
+
+        :rtype: str
+        """
+        return self._key
+
+    def update(self, key):
```

Comment:
Do a setter:
https://docs.python.org/3/library/functions.html#property

```python
class C:
    def __init__(self):
        self._x = None

    @property
    def x(self):
        """I'm the 'x' property."""
        return self._x

    @x.setter
    def x(self, value):
        self._x = value

    @x.deleter
    def x(self):
        del self._x
```
and not update

Reply:
Same issue was raised in .NET, but we decided on `update`. See @tg-msft [comment's](https://github.com/Azure/azure-sdk-for-net/pull/10668#discussion_r394458230):
"A property setter would work here, but not for a lot of other credential types. An Update method works better if you need to do any computation on the values, if you have multiple values that need to be swapped atomically, etc. This is more about following @schaabs's Credential pattern even though the Update method isn't as necessary with this simplest of credentials."

Reply:
"setter" in Python allows computation of values, this would look like this:
```python
    @property
    def key(self):
        # type () -> str
        """The value of the configured key.
        :rtype: str
        """
        return self._key

    @key.setter
    def key(self, key):
        # type: (str) -> None
        """Update the key.
        This can be used when you've regenerated your service key and want
        to update long-lived clients.
        :param str key: The key used to authenticate to an Azure service
        :raises: ValueError or TypeError
        """
        if not key:
            raise ValueError("The key used for updating can not be None or empty")
        if not isinstance(key, six.string_types):
            raise TypeError("The key used for updating must be a string.")
        self._key = key

Reply:
Would this be one of those scenarios where languages differ? @schaabs what do you think?

Reply:
No there should not be a setter but there should be an update method.  It doesn't really matter that much for this particular credential since it's just one property that makes up the key, but other credential types like this have multiple properties and so the purpose of the update method is to swap them in a thread safe way. I'd like to make this consistent with that pattern so credential types throughout our sdk provided a similar experience.

Reply:
I understand from what you say that in other credential, update might have more than one argument, then using an "update" consistently makes sense.

Reply:
Late to the game. Yes, we should use `update`. For exactly the reason(s) stated above. It is not the "normal" getter/setter situation where a property would absolutely be the right answer.